engine/generic_list.hpp

```diff
@@ -0,0 +1,777 @@
+    }
+  };
+
+ public:
+  // Default to an empty list with no name and 0 as id
+  // Must Init() or Restore() before further use.
+  GenericList() : Collection{"", CollectionIDType{}} {}
+
+  GenericList(GenericList const&) = delete;
+  GenericList(GenericList&&) = delete;
+  GenericList& operator=(GenericList const&) = delete;
+  GenericList& operator=(GenericList&&) = delete;
+  ~GenericList() = default;
+
+  // Initialize a List with pmem base address p_base, pre-space space,
+  // Creation time, List name and id.
+  void Init(AddressTranslator<DLRecord*> tran, SpaceEntry space,
+            TimeStampType timestamp, StringView const key,
+            CollectionIDType id) {
+    collection_name_.assign(key.data(), key.size());
+    collection_id_ = id;
+    atran = tran;
+    list_record = DLRecord::PersistDLRecord(
+        atran.addressOf(space.offset), space.size, timestamp,
+        RecordType::ListRecord, NullPMemOffset, NullPMemOffset, NullPMemOffset,
+        key, ID2String(id));
+  }
+
+  template <typename ListDeleter>
+  void Destroy(ListDeleter list_deleter) {
+    kvdk_assert(sz == 0 && list_record != nullptr && first == nullptr &&
+                    last == nullptr,
+                "Only initialized empty List can be destroyed!");
+    list_deleter(list_record);
+    list_record = nullptr;
+  }
+
+  bool Valid() const { return (list_record != nullptr); }
+
+  // Restore a List with its ListRecord, first and last element and size
+  // This function is used by GenericListBuilder to restore the List
+  void Restore(AddressTranslator<DLRecord*> tran, DLRecord* list_rec,
+               DLRecord* fi, DLRecord* la, size_t n) {
+    auto key = list_rec->Key();
+    collection_name_.assign(key.data(), key.size());
+    kvdk_assert(list_rec->Value().size() == sizeof(CollectionIDType), "");
+    collection_id_ = ExtractID(list_rec->Value());
+    atran = tran;
+    list_record = list_rec;
+    first = fi;
+    last = la;
+    sz = n;
+  }
+
+  LockType* Mutex() { return &mu; }
+
+  std::unique_lock<LockType> AcquireLock() {
+    return std::unique_lock<LockType>(mu);
+  }
+
+  void ExpireAt(ExpiredTimeType time) final {
+    list_record->PersistExpireTimeNT(time);
+    expire_time = time;
+  }
+
+  size_t Size() const { return sz; }
+
+  Iterator Front() { return ++Head(); }
+
+  Iterator Back() { return --Tail(); }
+
+  Iterator Head() const { return Iterator{this}; }
+
+  Iterator Tail() const { return Iterator{this}; }
+
+  Iterator Seek(std::int64_t index) {
+    if (index >= 0) {
+      auto iter = Front();
+      while (index != 0 && iter != Tail()) {
+        ++iter;
+        --index;
+      }
+      return iter;
+    } else {
+      auto iter = Back();
+      while (index != -1 && iter != Head()) {
+        --iter;
+        ++index;
+      }
+      return iter;
+    }
+  }
+
+  template <typename ElemDeleter>
+  Iterator Erase(Iterator pos, ElemDeleter elem_deleter) {
+    kvdk_assert(pos != Head(), "Cannot erase Head()");
```

Comment:
can erase Tail() ?

Reply:
Tail() is Head(). It's a dummy iterator. Please refer to comments.